test/libsolidity/TestHook.h

```diff
@@ -0,0 +1,52 @@
+/*
+	This file is part of solidity.
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "Builtin.h"
+#include <memory>
+#include <vector>
+
+namespace solidity::frontend::test
+{
+
+class TestFunctionCall;
+
+class TestHook
+{
+public:
+	TestHook() = default;
+	virtual ~TestHook() = default;
+
+	virtual void beginTestCase() {}
+	virtual void beforeFunctionCall([[maybe_unused]] TestFunctionCall const& _call) {}
+	virtual void afterFunctionCall([[maybe_unused]] TestFunctionCall const& _call) {}
+	virtual void endTestCase() {}
+
+	virtual bool verifyFunctionCall([[maybe_unused]] TestFunctionCall const& _call) { return true; }
+
+	virtual std::string formatFunctionCall(
+		[[maybe_unused]] const TestFunctionCall& _call,
```

Comment:
This does not really look nice. Do you need an implementation of this at all?

Reply:
Yes, @mijovic ask a similar question. I think it's better to have an implementation here like this, if not all "hooks" need to declare it in a similar way, especially if some of the functions don't do anything. So I think it's better to implement it here,  and inherited "hooks" can only implement a specific function if needed.

Reply:
Now `TestHook` is abstract.